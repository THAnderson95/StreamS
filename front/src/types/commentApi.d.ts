import { ATTACHMENT_TYPES } from "@/constants";

export interface APIDataCommentResponse {
  data: Array<CommentData>;
  paging: Paging;
}

interface CommentData {
  message: string;
  from: FromData;
  reactions?: Reactions;
  attachment?: Attachment;
  id: string;
  like_count: number;
  parent: Parent;
}

interface Parent {
  id: string;
}

interface FromData {
  id: string;
  name: string;
  picture: Picture;
}

interface Picture {
  data: PictureData;
}

interface PictureData {
  height: number;
  width: number;
  is_silhouette: boolean;
  url: string;
}

interface Attachment {
  media: {
    image: {
      height: number;
      src: string;
      width: number;
    };
    target: {
      id: string;
      url: string;
    };
    source?: string;
  };
  type: ATTACHMENT_TYPES;
  url: string;
}

interface Reactions {
  data: Array<Reaction>;
  paging: Paging;
}

interface Reaction {
  id: string;
  name: string;
  type: string;
  paging: Paging;
}

interface Paging {
  cursors: {
    before: string;
    after: string;
  };
}
