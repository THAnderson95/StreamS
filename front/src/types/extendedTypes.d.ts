import { CommentData } from "@api/commentApi";

export interface CommentDataWithReplies extends CommentData {
  replies?: Array<CommentData>;
}
