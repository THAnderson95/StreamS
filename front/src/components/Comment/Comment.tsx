import { useState } from "react";
import { COMMENT_TYPES } from "@/constants";
import { Attachment, CommentDataWithReplies } from "@/types";

interface CommentProps {
  data: CommentDataWithReplies;
}

export default function Comment({ data }: CommentProps) {
  const [collapseReplies, setCollapseReplies] = useState<boolean>(false);

  const renderAttachment = (attachment: Attachment) => {
    if (!attachment) {
      return null;
    }

    if (attachment.type === COMMENT_TYPES.ANIMATED_IMAGE_VIDEO) {
      return (
        <video
          src={attachment.media.source}
          className="p-2"
          autoPlay
          muted
          loop
        />
      );
    }

    return <img src={attachment.media.image.src} className="p-2" />;
  };

  return (
    <div className="comment-container border flex flex-row p-2 m-2 bg-gray-400 rounded-2xl">
      <div className="comment-picture">
        <img
          src={data.from.picture.data.url}
          className="comment-profile-image mr-4 rounded-full"
        ></img>
      </div>
      <div className="comment-info flex-1">
        <div className="comment-name text-2xl">{data.from.name}</div>
        <div className="comment-message italic">{data.message}</div>

        {data.attachment && (
          <div className="comment-attachment">{renderAttachment(data.attachment)}</div>
        )}
        {data.replies && data.replies.length > 0 && (
          <button
            className="show-button bg-blue-500 hover:bg-blue-600 text-white px-2 py-1 m-1 rounded"
            onClick={() => setCollapseReplies(!collapseReplies)}
          >
            {collapseReplies ? "Hide Replies..." : "Show Replies..."}
          </button>
        )}
        {collapseReplies &&
          data.replies?.map((reply) => <Comment key={reply.id} data={reply} />)}
      </div>
    </div>
  );
}
