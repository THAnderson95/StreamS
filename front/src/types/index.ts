export * from './commentApi';
export * from './extendedTypes'