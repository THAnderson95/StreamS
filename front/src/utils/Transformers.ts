import {
  APIDataCommentResponse,
  CommentData,
  CommentDataWithReplies,
} from "@/types";

export function addRepliesToComments(
  data: APIDataCommentResponse
): Array<CommentDataWithReplies> {
  // filter the comments that are child comments and convert them to `CommentDataWithReplies`
  const filteredComments: Array<CommentDataWithReplies> = data.data
    .filter((comment) => !comment.parent)
    .map((comment) => ({ ...comment, replies: [] }));

  // filter ONLY child replies
  const childReplies: Array<CommentData> = data.data.filter(
    (comment) => comment.parent
  );

  // merge them
  const addRepliesToComments: Array<CommentDataWithReplies> =
    filteredComments.map((parentComment) => {
      parentComment.replies = childReplies.filter(
        (replyComment) => replyComment.parent?.id === parentComment.id
      );
      return parentComment;
    });

  return addRepliesToComments;
}
