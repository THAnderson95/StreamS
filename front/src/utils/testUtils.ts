import { CommentDataWithReplies } from "@/types/extendedTypes";
import { ATTACHMENT_TYPES } from "@/constants";

function generateCommentWithoutParentsAndReplies(): CommentDataWithReplies {
  return {
    id: "1",
    message: "This is a comment without parents and without replies",
    from: {
      id: "user1",
      name: "User One",
      picture: {
        data: {
          height: 50,
          width: 50,
          is_silhouette: false,
          url: "https://example.com/user1.jpg",
        },
      },
    },
    like_count: 10,
    replies: [],
  };
}

function generateCommentWithParentAndReplies(): CommentDataWithReplies {
  return {
    id: "2",
    message: "This is a comment with a parent and with replies",
    from: {
      id: "user2",
      name: "User Two",
      picture: {
        data: {
          height: 50,
          width: 50,
          is_silhouette: false,
          url: "https://example.com/user2.jpg",
        },
      },
    },
    like_count: 5,
    parent: { id: "1" },
    attachment: undefined,
    reactions: undefined,
    replies: [
      {
        id: "3",
        message: "This is a reply to comment 2",
        from: {
          id: "user3",
          name: "User Three",
          picture: {
            data: {
              height: 50,
              width: 50,
              is_silhouette: false,
              url: "https://example.com/user3.jpg",
            },
          },
        },
        like_count: 2,
        parent: { id: "2" },
      },
    ],
  };
}

function generateCommentWithMultipleReactions(): CommentDataWithReplies {
  return {
    id: "4",
    message: "This is a comment with multiple reactions",
    from: {
      id: "user4",
      name: "User Four",
      picture: {
        data: {
          height: 50,
          width: 50,
          is_silhouette: false,
          url: "https://example.com/user4.jpg",
        },
      },
    },
    like_count: 20,
    reactions: {
      data: [
        { id: "reaction1", name: "User Five", type: "LIKE" },
        { id: "reaction2", name: "User Six", type: "LOVE" },
      ],
      paging: { cursors: { before: "", after: "" } },
    },
    replies: [],
  };
}

function generateCommentWithAttachment(): CommentDataWithReplies {
  return {
    id: "5",
    message: "",
    from: {
      id: "user5",
      name: "User Five",
      picture: {
        data: {
          height: 50,
          width: 50,
          is_silhouette: false,
          url: "https://example.com/user5.jpg",
        },
      },
    },
    like_count: 0,
    attachment: {
      media: {
        image: {
          height: 100,
          src: "https://example.com/attachment.jpg",
          width: 100,
        },
        target: { id: "attachment1", url: "https://example.com/attachment" },
      },
      type: ATTACHMENT_TYPES.PHOTO,
      url: "https://example.com/attachment",
    },
    replies: [],
  };
}

function generateCommentWithOnlyMessage(): CommentDataWithReplies {
  return {
    id: "6",
    message: "This is a comment with only a message",
    from: {
      id: "user6",
      name: "User Six",
      picture: {
        data: {
          height: 50,
          width: 50,
          is_silhouette: false,
          url: "https://example.com/user6.jpg",
        },
      },
    },
    like_count: 0,
    replies: [],
  };
}

export function generateCommentData(): Array<CommentDataWithReplies> {
  return [
    generateCommentWithoutParentsAndReplies(),
    generateCommentWithParentAndReplies(),
    generateCommentWithMultipleReactions(),
    generateCommentWithAttachment(),
    generateCommentWithOnlyMessage(),
  ];
}