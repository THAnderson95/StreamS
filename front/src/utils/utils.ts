import { REACTION_TYPE_SYMBOLS, REACTION_TYPES } from "@/constants";

export function getReactionSymbol(reaction: REACTION_TYPES): string {
    return REACTION_TYPE_SYMBOLS[reaction]
}