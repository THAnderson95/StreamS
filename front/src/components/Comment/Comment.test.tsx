import { act } from "react"; // This should be from 'react' now
import { render } from "@testing-library/react";

import Comment from "./Comment";
import { CommentDataWithReplies } from "@/types/extendedTypes";
import { generateCommentData } from "@/utils/testUtils";

it("renders correctly and matches snapshot", async () => {
  const commentData: CommentDataWithReplies = generateCommentData()[0];
  const { asFragment } = render(
    <Comment key={commentData.id} data={commentData} />,
  );

  /* eslint-disable testing-library/no-unnecessary-act */
  await act(async () => {});

  expect(asFragment()).toMatchSnapshot();
});
