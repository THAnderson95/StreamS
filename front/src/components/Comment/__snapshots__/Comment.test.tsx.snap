// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="comment-container border border-2 flex flex-row p-2 m-2 bg-gray-300 rounded-2xl"
  >
    <div
      class="comment-picture"
    >
      <img
        alt="User One"
        class="comment-profile-image mr-4 rounded-full max-w-[50px] max-h-[50px]"
        src="https://example.com/user1.jpg"
      />
    </div>
    <div
      class="comment-info flex-1"
    >
      <div
        class="content-container bg-gray-200 rounded-lg m-2 p-2"
      >
        <div
          class="comment-name font-bold text-lg"
        >
          User One
        </div>
        <div
          class="comment-message"
        >
          This is a comment without parents and without replies
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
